### PR TITLE
Add HTMLFormatter in formatter.exs templates

### LINKED ISSUE #4749: Have phx.new automatically include heex formatting
It could be nice to have `mix phx.new` include the necessary changes to `.formatter.exs` to format `.heex` files upon running `mix format`. 

```elixir
# .formatter.exs
[
  plugins: [Phoenix.LiveView.HTMLFormatter],
  import_deps: [:ecto, :phoenix],
  inputs: ["*.{heex,ex,exs}", "priv/*/seeds.exs", "{config,lib,test}/**/*.{heex,ex,exs}"],
  subdirectories: ["priv/*/migrations"]
]
```

Are there any considerations to keep in mind if we were to implement this?

### PR DESCRIPTION
**Pull request by leandrocp**
That's an attempt to close #4749 

The config is added only if `@html` and Elixir >= 1.13.4 to meet formatter requirements.

### CODE REVIEW

**Comment by josevalim** [comment]
According to this commit we also need to add `{:phoenix_live_view, ">= 0.0.0"}` at the umbrella root with a comment that it is used exclusively by the formatter: https://github.com/phoenixframework/phoenix_live_view/commit/9793894a7da7588c1a3913da0a687b707aa14b98

Can you please do so? 

Thank you!

**Comment by leandrocp** [comment]
> According to this commit we also need to add `{:phoenix_live_view, ">= 0.0.0"}` at the umbrella root with a comment that it is used exclusively by the formatter: [phoenixframework/phoenix_live_view@9793894](https://github.com/phoenixframework/phoenix_live_view/commit/9793894a7da7588c1a3913da0a687b707aa14b98)
> 
> Can you please do so?
> 
> Thank you!

Sure!

File: installer/templates/phx_umbrella/mix.exs [hunk 1]
```diff
@@ -23,8 +23,10 @@ defmodule <%= @root_app_module %>.MixProject do
   #
   # Dependencies listed here are available only for this project
   # and cannot be accessed from applications inside the apps/ folder.
-  defp deps do
-    []
+  defp deps do<%= if @html and Version.match?(System.version(), ">= 1.13.4") do %>
+    {:phoenix_live_view, ">= 0.0.0"} # required by Phoenix.LiveView.HTMLFormatter
```
Comment (josevalim) [inline installer/templates/phx_umbrella/mix.exs]:
```suggestion
    [
      # Required to run "mix format" on ~H/.heex files from the umbrella root 
      {:phoenix_live_view, ">= 0.0.0"} 
    ]
```

File: installer/templates/phx_umbrella/mix.exs [hunk 2]
```diff
@@ -23,8 +23,10 @@ defmodule <%= @root_app_module %>.MixProject do
...
+<% else %>
+    []<% end %>
   end
```
Comment (josevalim) [inline installer/templates/phx_umbrella/mix.exs:31]:
```suggestion
  defp deps do<%= if @html and Version.match?(System.version(), ">= 1.13.4") do %>
    [
      # Required to run "mix format" on ~H/.heex files from the umbrella root 
      {:phoenix_live_view, ">= 0.0.0"} 
    ]<% else %>
    []<% end %>
  end
```

File: installer/test/phx_new_umbrella_test.exs [hunk 3]
```diff
@@ -45,6 +45,17 @@ defmodule Mix.Tasks.Phx.New.UmbrellaTest do
         assert file =~ "apps_path: \"apps\""
       end
 
+      # Phoenix.LiveView.HTMLFormatter
+      if Version.match?(System.version(), ">= 1.13.4") do
+        assert_file root_path(@app, "mix.exs"), fn file ->
+          assert file =~ "defp deps do\n    {:phoenix_live_view, \">= 0.0.0\"}"
```
Comment (josevalim) [inline installer/test/phx_new_umbrella_test.exs]:
```suggestion
          assert file =~ "defp deps do\n    [\n      {:phoenix_live_view, \">= 0.0.0\"}"
```

File: installer/templates/phx_umbrella/mix.exs [hunk 4]
```diff
@@ -23,8 +23,13 @@ defmodule <%= @root_app_module %>.MixProject do
   #
   # Dependencies listed here are available only for this project
   # and cannot be accessed from applications inside the apps/ folder.
-  defp deps do
-    []
+  defp deps do<%= if @html and Version.match?(System.version(), ">= 1.13.4") do %>
+  defp deps do<%= if @html and Version.match?(System.version(), ">= 1.13.4") do %>
```
Comment (josevalim) [inline installer/templates/phx_umbrella/mix.exs:27]:
```suggestion
```

File: installer/test/phx_new_umbrella_test.exs [hunk 5]
```diff
@@ -48,7 +48,7 @@ defmodule Mix.Tasks.Phx.New.UmbrellaTest do
       # Phoenix.LiveView.HTMLFormatter
       if Version.match?(System.version(), ">= 1.13.4") do
         assert_file root_path(@app, "mix.exs"), fn file ->
-          assert file =~ "{:phoenix_live_view, \">= 0.0.0\"}"
+          assert file =~ ~r/defp deps do\n.*\[\n.*\n.*{:phoenix_live_view, \">= 0.0.0\"}/
```
Comment (leandrocp) [inline installer/test/phx_new_umbrella_test.exs:51]: @josevalim both are fine, it's up to you if should be more specific of if just checking `{:phoenix_live_view, \">= 0.0.0\"}"` is enough

Comment (josevalim) [reply installer/test/phx_new_umbrella_test.exs:51]: I think just checking the tuple is enough!

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Comment by michallepicki** [comment]
For the plugin to work as expected in an umbrella project, the plugin needs to be also added to the top level `formatter.exs` file in umbrella template, as shown in [this comment](https://github.com/phoenixframework/phoenix_live_view/issues/1969#issuecomment-1100673177=)

**Comment by leandrocp** [comment]
> For the plugin to work as expected in an umbrella project, the plugin needs to be also added to the top level `formatter.exs` file in umbrella template, as shown in [this comment](https://github.com/phoenixframework/phoenix_live_view/issues/1969#issuecomment-1100673177)

Thanks @michallepicki just pushed a PR https://github.com/phoenixframework/phoenix/pull/4800

**Comment by chrismccord** [comment]
@leandrocp can you check the integration tests on master? I get failures locally on our tests that ensure the generated code is formatted. It fails on layout files and phx.gen.html files, because presumably they have minor tweaks with the formatter. Thanks!

**Comment by leandrocp** [comment]
> @leandrocp can you check the integration tests on master? I get failures locally on our tests that ensure the generated code is formatted. It fails on layout files and phx.gen.html files, because presumably they have minor tweaks with the formatter. Thanks!

Sure. I can reproduce, working on it.

**Comment by leandrocp** [comment]
@chrismccord one of the changes is adding parens to template function, eg: https://github.com/phoenixframework/phoenix/blob/dcc7ae530767896d458fcfde95156e037f3adc36/installer/templates/phx_web/templates/layout/root.html.heex#L8 is generated and formatted as:

```heex
    <%= live_title_tag(assigns[:page_title] || "app", suffix: " · Phoenix Framework") %>
```

I'm wondering if such functions should be added to `locals_without_parens` in Phoenix's default `.formatter.exs`?

**Comment by chrismccord** [comment]
parens is fine. We'll probably convert this to a function component for v1.7. Thanks!

**Pull request merged**